config/reader.go

```diff
@@ -390,24 +390,36 @@ func processConfigLocalConfig(unprocessedConfig *Config, logger logging.Logger)
 	return processConfig(unprocessedConfig, false, logger)
 }
 
+// processConfig processes the config passed in. The config can be either JSON or gRPC derived.
+// If any part of this function errors, the function will exit and no part of the new config will be returned
+// until it is corrected.
 func processConfig(unprocessedConfig *Config, fromCloud bool, logger logging.Logger) (*Config, error) {
+	// Ensure validates the config but also substitutes in some defaults. Implicit dependencies for builtin resource
+	// models are not filled in until attributes are converted.
 	if err := unprocessedConfig.Ensure(fromCloud, logger); err != nil {
 		return nil, err
 	}
 
+	// we cached the unprocessed config, so make a copy before changing it too much. Also ensures validation
+	// happens again on resources, remotes, and modules.
 	cfg, err := unprocessedConfig.CopyOnlyPublicFields()
 	if err != nil {
 		return nil, errors.Wrap(err, "error copying config")
 	}
 
+	// Copy does not preserve ConfigFilePath since it preserves only JSON-exported fields and so we need
+	// to pass it along manually. ConfigFilePath needs to be preserved so the correct config watcher can
+	// be instantiated later in the flow.
+	cfg.ConfigFilePath = unprocessedConfig.ConfigFilePath
+
+	// replacement can happen in resource attributes and in the module config. look at config/placeholder_replace.go
+	// for available substitution types.
 	if err := cfg.ReplacePlaceholders(); err != nil {
 		logger.Errorw("error during placeholder replacement", "err", err)
 	}
 
-	// Copy does not presve ConfigFilePath and we need to pass it along manually
-	cfg.ConfigFilePath = unprocessedConfig.ConfigFilePath
-
-	// See if default service already exists in the config
+	// See if default service already exists in the config and add them in if not. This code allows for default services to be
+	// defined under a name other than "builtin".
```

Comment:
Ah neat, I don't think I picked up the relationship between default services and the `builtin` keyword.